feat: Tray icon voice detection

PolisanTheEasyNick opened the pull request:
epic pr🚀 cooking explodin
Resolves #294, partially resolves #667
- [x] Show icons depending on current voice state

Any ideas about improving PR structure are welcome ^^

---

### Voice Detection Icons
![image](https://github.com/user-attachments/assets/514337f5-9614-4756-bc07-550e11ba763f) ![image](https://github.com/user-attachments/assets/555d6198-90ac-4819-9cef-3c0f6ce09a80) ![image](https://github.com/user-attachments/assets/ac04f42c-e6a7-47ac-9f58-1482a53c302f) ![image](https://github.com/user-attachments/assets/9301bbc3-e7e3-45b8-aefa-99aaaef5504a)



---

### Notification Badge on top of Tray Icon
![image](https://github.com/user-attachments/assets/eac5978d-2ec0-4bf5-ba68-1f5eee445fe3)

Special thanks to @MrGarlic1 and @Covkie 

Also, on my repo, there is branch `tray-icon-old` where all changes with tray icons customizations, colors, and so on

PolisanTheEasyNick committed:
Added new tray icons

PolisanTheEasyNick committed:
Added tray icon voice detection

PolisanTheEasyNick committed:
removed unused imports

PolisanTheEasyNick committed:
Fixed for tests

PolisanTheEasyNick committed:
Removed unused logger

PolisanTheEasyNick committed:
Fixed icon when connected with muted status

PolisanTheEasyNick committed:
added ability to change icon to unread

PolisanTheEasyNick committed:
Revert "added ability to change icon to unread"

This reverts commit 3ccb0c40fc5255aa9e86f84df62bd42cb2037e65.

PolisanTheEasyNick committed:
Increased border width

PolisanTheEasyNick commented:
Thank you for feedback! I'm not using any black colors, I believe that it was just small border width. Also, I increased whole icon size. Can you please pull and retest?
What state is on your last screenshot? If it is muted or deafened that means that icons too small, and we need to redraw them completely, because I just took discord's original icons and changed colors to Vesktop ones. They are 24x24 pixels, so maybe they have not resized to proper size in Gnome system.

EtereosDawn commented:
> Thank you for feedback! I'm not using any black colors, I believe that it was just small border width. Also, I increased whole icon size. Can you please pull and retest? What state is on your last screenshot? If it is muted or deafened that means that icons too small, and we need to redraw them completely, because I just took discord's original icons and changed colors to Vesktop ones. They are 24x24 pixels, so maybe they have not resized to proper size in Gnome system.

I can confirm that it looks much better! Well, maybe the edges of the circle could be a little smoother but I believe that the pixels are due to my monitor. That's great :) thanks for accepting the feedbeck.

![image](https://github.com/Vencord/Vesktop/assets/34200212/dccaf1d1-fc36-4205-89a8-ca2564afd03a)

Vendicated commented on src/renderer/patches/tray.ts:
```diff
@@ -0,0 +1,63 @@
+                VesktopNative.app.setTrayIcon("muted");
+            } else {
+                VesktopNative.app.setTrayIcon("idle");
+            }
```
this if, else if, else block is duplicated 4 times. it should be extracted into a function

Vendicated commented on src/main/mainWindow.ts:
```diff
@@ -478,3 +478,23 @@ export async function createWindows() {
     initArRPC();
 }
+
+export async function setTrayIcon(iconName) {
```
```suggestion
export async function setTrayIcon(iconName: string) {
```

Vendicated commented on src/main/mainWindow.ts:
```diff
@@ -478,3 +478,23 @@ export async function createWindows() {
+    if (!tray) return;
+    switch (iconName) {
+        case "speaking":
+            tray.setImage(SPEAKING_ICON_PATH);
```
whats the point of making all these constants? why not just
```js
const Icons = new Set(["speaking", "muted", "deafened", "idle", "main"]);
// ...

if (!Icons.has(iconName)) {
	console.warn("setTrayIcon: Invalid icon name", iconName);
	iconName = "main";
}
tray.setImage(join(ICON_DIR, iconName + ".png"));
```

PolisanTheEasyNick committed:
Updated and resized icons

PolisanTheEasyNick committed:
changes according to reviews

PolisanTheEasyNick committed:
fixed bug with not changing to default icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Added custom color support

PolisanTheEasyNick commented:
Added ability to change colors. For now, I want to find a way how to disable icon color chooser in settings when tray icon is disabled instead of hiding the whole setting and (maybe) add ability to propose to user his system accent color.

Vendicated commented on src/renderer/components/settings/TrayColorPicker.tsx:
```diff
@@ -0,0 +1,46 @@
+    "#F6BFAC" // Vesktop inpired
+];
+
+export const trayIconPicker: SettingsComponent = ({ settings }) => {
```
react components should be PascalCase

Vendicated commented on src/renderer/components/settings/TrayColorPicker.tsx:
```diff
@@ -0,0 +1,46 @@
+    return (
+        <div className="tray-settings">
+            <div className="tray-container">
+                <div className="tray-settings-labels">
```
we prefix our classes to prevent potential conflicts

vencord usually uses `.vc-*` whereas vesktop usually uses `.vcd-`

Vendicated commented on src/renderer/components/settings/TrayColorPicker.tsx:
```diff
@@ -0,0 +1,46 @@
+];
+
+export const trayIconPicker: SettingsComponent = ({ settings }) => {
+    if (!settings.tray) return null; // how to disable instead of hiding?
```
is this actually necessary to change? returning null is the correct way in react to render nothing

PolisanTheEasyNick replied:
But in settings we also have "Minimize to tray" which disables when Tray is off and I wanted to achieve similar behavior for this too. Because one element just *grays* as inactive and another (mine) disappearing at all

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
resolved due to reviews

PolisanTheEasyNick committed:
fixed case in Settings.tsx

PolisanTheEasyNick committed:
svg rework, add proposal for accent color (windows,linux), some bug fixes

PolisanTheEasyNick committed:
add more colors to presets

PolisanTheEasyNick committed:
fixed check for tray availability

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
fixed tray updating in minimized window at Windows. This also fixes random disconnects while window minimized.

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'Vencord:main' into tray-icon

PolisanTheEasyNick committed:
Removed backgroundThrottling setting

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
use one voiceActions for both mute and deaf

PolisanTheEasyNick committed:
Merge branch 'Vencord:main' into tray-icon

PolisanTheEasyNick committed:
fix: show proper icon after toggle setting in VC

PolisanTheEasyNick committed:
svg: change fill color depending on theme

PolisanTheEasyNick committed:
feat: ablity to manually set fill color

PolisanTheEasyNick committed:
feat: save icons into vesktop cache

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
feat: ability to disable icons rewriting

PolisanTheEasyNick committed:
fix: force regenerate icons if can't open

PolisanTheEasyNick committed:
refactor: use paths instead of filenames

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

Commit:
fix: add pencil icon/tray setting css classes

Commit:
ux: condense customize tray settings

PolisanTheEasyNick committed:
ui: modal for picking custom tray icons

PolisanTheEasyNick committed:
ui: fix delimeters

PolisanTheEasyNick committed:
feat: ability to set custom icons

PolisanTheEasyNick committed:
mac: resize icons to 16x16

PolisanTheEasyNick committed:
feat: overrides by each icon type

PolisanTheEasyNick committed:
ui: changed to fit Discord

PolisanTheEasyNick committed:
ui: icon namings at right side

PolisanTheEasyNick committed:
feat: notification badge at tray

PolisanTheEasyNick committed:
feat: custom badges colors

PolisanTheEasyNick committed:
fix: check for luminance only for badge

PolisanTheEasyNick committed:
fix: reset button

PolisanTheEasyNick committed:
feat: do not show count of notifs

PolisanTheEasyNick committed:
fix: show badge only for unreads

PolisanTheEasyNick committed:
feat: accent color support for linix

PolisanTheEasyNick committed:
feat: choose icon from .svg

PolisanTheEasyNick committed:
fix: generate only one icon on error

PolisanTheEasyNick committed:
css: remove unused classes

PolisanTheEasyNick committed:
some clear-ups

PolisanTheEasyNick committed:
fix: default tray when stream stopped

PolisanTheEasyNick committed:
fix: speak icon when stream with sound started

PolisanTheEasyNick commented:
> displays the speaking/audio detected icon despite no user-speaking occurring.

Fixed in latest commit)
Interesting thing is that Discord calls SPEAKING event even when sound outcoming from video stream, so I've added additional checks for event call context.

D3SOX commented:
Great work! Can you add an option to disable the badge or have it only for pings? I find it very annoying because I pretty much always have some unread messages

PolisanTheEasyNick committed:
badge: show mentions instead of unreads

PolisanTheEasyNick committed:
fix: proper follow "Notification Badge" setting

Co-authored-by: Cookie <52550063+Covkie@users.noreply.github.com>

Covkie committed:
Reword and restructure tray settings

D3SOX commented on src/main/tray.ts:
```diff
@@ -0,0 +1,187 @@
+    if (!Icons.has(iconName)) return;
+
+    // if need to set main icon then check whether there is need of notif badge
+    if (iconName === "icon" && lastBadgeCount > 0) {
```
I'm not sure if I understand the new behavior correctly. After updating, it now shows the dot for non-mentions again.
Am I supposed to disable “Enable Unread Message Badge” when I want it to only for mentions?
![image](https://github.com/Vencord/Vesktop/assets/24937357/502ad68b-0583-463f-be67-db42300ab810)

I think this could be made clearer

PolisanTheEasyNick committed:
feat: dynamic change tray bg color on theme change

Vendicated committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
fixes for vesktop update

PolisanTheEasyNick committed:
feat: add ability to use dynamic accent color

PolisanTheEasyNick committed:
fix: add setting

PolisanTheEasyNick committed:
fix: accent color for badge

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

alchzh commented:
> Might it be best to fold #261 into this PR?

I just found out about this PR, looks great though I haven't tested it yet. This only shows a notification dot not a number, right? if there's interest in getting numbers in the tray icons like I have it in that PR i'm happy to work on top of your PR @PolisanTheEasyNick (it's an important feature for me)

PolisanTheEasyNick commented:
> > Might it be best to fold #261 into this PR?
> 
> I just found out about this PR, looks great though I haven't tested it yet. This only shows a notification dot not a number, right? if there's interest in getting numbers in the tray icons like I have it in that PR i'm happy to work on top of your PR @PolisanTheEasyNick (it's an important feature for me)

I can add that easily (even already did once at [c725352](https://github.com/Vencord/Vesktop/pull/517/commits/c725352bef0dd85f110956593e8a5a79a540d900) and removed at [8bbe16b](https://github.com/Vencord/Vesktop/pull/517/commits/8bbe16b775a6b5b9b2ac6fd8ec168a6cf69fc8e5)), but problem was that numbers on badge in tray icon is very small and not readable, so there was a decision in ds chat to show only notification dot.
Also, this PR is already quite massive and badge logic on Linux is quite overwhelming, so probably the best approach here will be to rework the whole notification badge settings and with this add support of enabling/disabling badge numbers in tray icon :)

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
ref: use dbus-native for getting accent color

PolisanTheEasyNick committed:
fix: update pnpm-lock

PolisanTheEasyNick committed:
fix exploded tests

zoobporsor commented:
Where is the option to change the taskbar and tray icon? I cant find it

PolisanTheEasyNick commented:
> Where is the option to change the taskbar and tray icon? I cant find it

Vesktop Settings -> Tray  
![image](https://github.com/user-attachments/assets/ba239c4a-b783-4807-b98a-fb56832a297d)
If you still can't find try retoggle "Enable Tray Icon" option. If still can't find then make sure you're using `tray-icon` branch

PolisanTheEasyNick committed:
fix: show custom icon on startup initially

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

prochy-exe commented:
Hello, i tried building your branch for vencord however i get and error endlessly and the icon doesnt change what so ever.
![Vesktop_2024-08-27_14-06-14](https://github.com/user-attachments/assets/e85ab621-c625-4451-a7db-972ae0c0b17d)

Apart from that this is also being spammed in the console. That's all the useful info i could find.
![region_2024-08-27_14-07-37](https://github.com/user-attachments/assets/14754d0c-887d-4d8a-9d42-de81a313f4bb)

I am running this on arch, running hyprland.

PolisanTheEasyNick commented:
> I tried running this on windows and it works just fine, the only problem is that the icon is not customizable 

Try retoggle "Enable tray icon setting"   

> The only problem is that it looks a bit strange? I'm not sure if this is visible on the screenshot

It is strange indeed. Looks like some problem with icon sizes on Windows :c

zoobporsor commented:
> > Where is the option to change the taskbar and tray icon? I cant find it
> 
> Vesktop Settings -> Tray ![image](https://private-user-images.githubusercontent.com/39007846/360969720-ba239c4a-b783-4807-b98a-fb56832a297d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyOTQyODAsIm5iZiI6MTcyNTI5Mzk4MCwicGF0aCI6Ii8zOTAwNzg0Ni8zNjA5Njk3MjAtYmEyMzljNGEtYjc4My00ODA3LWI5OGEtZmI1NjgzMmEyOTdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDE2MTk0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiZTA2ZWQ5ODAwZGM4MDAyZjAxZWI1OTk1OTdhNjJhMmI5NmJiNzIxM2ExNjhiMzgxOGU1MWZiY2E0YjdmOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NGkDbVEvUHc2eCNLepRI0P7Ut0uMS-PY_oL1IOjsyaY) If you still can't find try retoggle "Enable Tray Icon" option. If still can't find then make sure you're using `tray-icon` branch

Ok but this doesnt exist in the main .app image? I need to download some special pre-release build?  Its not out yet officially?

PolisanTheEasyNick commented:
> Ok but this doesnt exist in the main .app image? 

No, it will be at .app when this PR would be merged and new Vesktop release released
> I need to download some special pre-release build? Its not out yet officially?

You can clone `tray-icon` branch from my fork and start Vesktop from console

prochy-exe commented:
> > > Where is the option to change the taskbar and tray icon? I cant find it
> > 
> > 
> > Vesktop Settings -> Tray ![image](https://private-user-images.githubusercontent.com/39007846/360969720-ba239c4a-b783-4807-b98a-fb56832a297d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyOTQyODAsIm5iZiI6MTcyNTI5Mzk4MCwicGF0aCI6Ii8zOTAwNzg0Ni8zNjA5Njk3MjAtYmEyMzljNGEtYjc4My00ODA3LWI5OGEtZmI1NjgzMmEyOTdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDE2MTk0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiZTA2ZWQ5ODAwZGM4MDAyZjAxZWI1OTk1OTdhNjJhMmI5NmJiNzIxM2ExNjhiMzgxOGU1MWZiY2E0YjdmOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NGkDbVEvUHc2eCNLepRI0P7Ut0uMS-PY_oL1IOjsyaY) If you still can't find try retoggle "Enable Tray Icon" option. If still can't find then make sure you're using `tray-icon` branch
> 
> Ok but this doesnt exist in the main .app image? I need to download some special pre-release build? Its not out yet officially?

i made this repo so if you trust me enough you can get it from here, if you are not on mac that is https://github.com/prochy-exe/custom-vesktop-builds

BreadYogurt commented:
Using SVG files doesn't seem to scale properly for me, here's the results from trying to use the discord tray icons included in the Papirus icon set.
![Screenshot_20240903_093546](https://github.com/user-attachments/assets/19bd0a87-f41f-46e0-afb2-b4fa63275bcc)

PolisanTheEasyNick commented:
Can you please send .svg file you're trying to set? I'll try to test and fix ASAP

BreadYogurt commented:
Here it is.
![discord-tray](https://github.com/user-attachments/assets/44052741-c7db-4df8-8a19-b19e60ff3634)

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
fix: svg scaling

BreadYogurt commented:
Yep, that fixed it, and more importantly, nothing exploded!

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
fix: finding color picker component

PolisanTheEasyNick commented on src/renderer/components/settings/TraySettings.tsx:
```diff
@@ -0,0 +1,269 @@
+
+import { SettingsComponent } from "./Settings";
+
+const ColorPicker = findByCodeLazy(".Messages.USER_SETTINGS_PROFILE_COLOR_SELECT_COLOR", ".BACKGROUND_PRIMARY)");
```
Fixed in 4bd812a, thanks @Covkie!

PolisanTheEasyNick committed:
Merge remote-tracking branch 'origin/main' into tray-icon

PolisanTheEasyNick committed:
fix pnpm lock

PolisanTheEasyNick committed:
headers fixes

PolisanTheEasyNick committed:
reimplenet edit-button

mrbigmouth502 commented:
The lack of a system tray notification dot in Vencord/Vesktop is the #1 thing keeping me from switching to it from the stock Discord client. It's an extremely important feature for my workflow. 

Hopefully the conflicts preventing it from being merged can be sorted out. 🙂

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
update pnpm-lock

Thespikedballofdoom commented:
> The lack of a system tray notification dot in Vencord/Vesktop is the number one thing keeping me from switching to it from the stock Discord client. It's an extremely important feature for my workflow.
> 
> Hopefully the conflicts preventing it from being merged can be sorted out. 🙂

Same, really, really pining for this, it's the only thing I look for every update.

PolisanTheEasyNick commented:
Thank you for response!

> You seem to have added a native dbus library. Do we really need this? 

I'm using it for getting accent color from Linux system. The library was added because it may be used for some other features, like #686.  
Personally, I think that feature "automatically pick accent color and propose it to user" is not very critical, because user can just pick color by himself, so I will remove library and accent color feature from this PR scope.

> but remove the settings ui

Should I remain color picker settings?

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
add dbus-native after merging main

PolisanTheEasyNick committed:
dbus-native fixes + colorpicker fix

KarlOfDuty commented:
> I do plan to have one central UI for managing various images but that should be part of a separate change

Is this planned soon? If its like a year away couldn't this gui be allowed to exist until then? It would be a shame to have these features ready to go but then get removed until really far into the future.

PolisanTheEasyNick committed:
removed color and icon change logic

PolisanTheEasyNick committed:
removed old svgs

PolisanTheEasyNick committed:
remove baking badge into icon logic

PolisanTheEasyNick committed:
remove dbus-native deps

PolisanTheEasyNick committed:
add line at end of file

PolisanTheEasyNick committed:
sync pnpm-lock

PolisanTheEasyNick committed:
minor end-fixes

Vendicated committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon

PolisanTheEasyNick committed:
tray icon changing functionality

PolisanTheEasyNick committed:
show notification badge after call

PolisanTheEasyNick committed:
final cleanups

PolisanTheEasyNick committed:
remove old icons

PolisanTheEasyNick committed:
use let instead of var

PolisanTheEasyNick committed:
Merge branch 'main' into tray-icon
